metals/src/main/scala/scala/meta/internal/metals/ammonite/Ammonite.scala

```diff
@@ -0,0 +1,470 @@
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier
+import org.eclipse.lsp4j.CompletionList
+import org.eclipse.lsp4j.Hover
+import org.eclipse.lsp4j.Position
+import org.eclipse.lsp4j.{Range => LspRange}
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.ExecutionContextExecutorService
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.meta.inputs.Input
+import scala.meta.internal.builds.BuildTools
+import scala.meta.internal.metals._
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+import scala.util.control.NonFatal
+import scala.util.Failure
+import scala.util.Success
+
+final class Ammonite(
+    buffers: Buffers,
+    compilers: Compilers,
+    compilations: Compilations,
+    statusBar: StatusBar,
+    diagnostics: Diagnostics,
+    doctor: Doctor,
+    tables: () => Tables,
+    languageClient: MetalsLanguageClient,
+    buildClient: ForwardingMetalsBuildClient,
+    userConfig: () => UserConfiguration,
+    indexWorkspace: () => Future[Unit],
+    workspace: () => AbsolutePath,
+    focusedDocument: () => Option[AbsolutePath],
+    buildTargets: BuildTargets,
+    buildTools: () => BuildTools,
+    config: MetalsServerConfig
+)(implicit ec: ExecutionContextExecutorService)
+    extends Cancelable {
+
+  def buildServer: Option[BuildServerConnection] =
+    buildServer0
+  def lastImportedBuild: ImportedBuild =
+    lastImportedBuild0
+
+  private var buildServer0 = Option.empty[BuildServerConnection]
+  private var lastImportedBuild0 = ImportedBuild.empty
+  private var lastImportVersions = VersionsOption(None, None)
+
+  private val cancelables = new MutableCancelable()
+  private val isCancelled = new AtomicBoolean(false)
+  def cancel(): Unit = {
+    if (isCancelled.compareAndSet(false, true)) {
+      val buildShutdown = buildServer0 match {
+        case Some(build) => build.shutdown()
+        case None => Future.unit
+      }
+      try cancelables.cancel()
+      catch {
+        case NonFatal(_) =>
+      }
+      try buildShutdown.asJava.get(100, TimeUnit.MILLISECONDS)
+      catch {
+        case _: TimeoutException =>
+      }
+    }
+  }
+
+  def loaded(path: AbsolutePath): Boolean =
+    path.isAmmoniteScript && {
+      val uri = path.toNIO.toUri.toASCIIString
+      lastImportedBuild0.targetUris.contains(uri)
+    }
+
+  def importBuild(): Future[Unit] =
+    buildServer0 match {
+      case None =>
+        Future.failed(new Exception("No Ammonite build server running"))
+      case Some(buildServer) =>
+        compilers.cancel()
+        for {
+          build0 <- statusBar.trackFuture(
+            "Importing Ammonite scripts",
+            MetalsLanguageServer.importedBuild(buildServer)
+          )
+          _ = {
+            lastImportedBuild0 = build0
+          }
+          _ <- indexWorkspace()
+          _ <- Future.sequence[Unit, List](
+            compilations
+              .cascadeCompileFiles(buffers.open.toSeq) ::
```

Comment:
Should we only compile ammonite scripts here?

Reply:
Probably, yes. Note that the whole workspace (Ammonite scripts + main project) is re-indexed, even though we don't re-import the build for the main project, we just re-use what was already imported. I'm not sure if we need to re-compile the main project or not, in that case.

Reply:
Since, nothing changes for the Bloop connection, no recompilation should be needed.